### PR TITLE
Rewrite concat

### PR DESCRIPTION
**Pull request by natebosch**
Uses sync forwarders for dart-lang/tools#1754

The simple implementation leaves out a lot of edge cases around
single-subscription vs broadcast streams and how they combine. Rewrite
to handle pause, resume, and cancel manually.

- Loop over stream types in tests
- Add test for onDone and canceling and relistening
- Handle mix of stream types by choosign whether to cancel or pause as
  appropriate.

### CODE REVIEW

File: lib/src/concat.dart [hunk 1]
```diff
@@ -17,11 +20,66 @@ class _Concat<T> implements StreamTransformer<T, T> {
...
+      firstDone = true;
+      currentStream = _next;
+      currentDoneHandler = onSecondDone;
+      listen();
+    }
+
+    currentDoneHandler = onFirstDone;
+
+    controller.onListen = () {
+      if (firstDone &&
```
Comment (jakemac53) [inline lib/src/concat.dart]: I am a bit confused by a lot of the logic here, why do we only care if `_next` is a broadcast stream, shouldn't most of these be checking for `currentStream.isBroadcast`?

File: test/concat_test.dart [hunk 2]
```diff
@@ -8,40 +8,121 @@ import 'package:test/test.dart';
...
+        StreamSubscription subscription;
+
+        setUp(() async {
+          firstCanceled = false;
+          secondCanceled = false;
+          secondListened = false;
+          first = streamTypes[firstType]()
+            ..onCancel = () {
+              firstCanceled = true;
+            };
```
Comment (jakemac53) [inline test/concat_test.dart:37]: should we be tracking `onListen` for this as well?

File: test/concat_test.dart [hunk 3]
```diff
@@ -8,40 +8,121 @@ import 'package:test/test.dart';
...
+        });
+
+        test('Does not listen to second stream before first stream finishes',
+            () async {
+          expect(secondListened, false);
+          await first.close();
+          expect(secondListened, true);
+        });
+
+        test('cancels single-subscription first stream on cancel', () async {
```
Comment (jakemac53) [inline test/concat_test.dart]: Should this be inside an `if (firstType == 'single-subscription')` block (the test name indicates that at least).

File: test/concat_test.dart [hunk 4]
```diff
@@ -8,40 +8,121 @@ import 'package:test/test.dart';
...
+          await subscription.cancel();
+          expect(firstCanceled, true);
+        });
+
+        if (firstType == 'single-subscription') {
```
Comment (jakemac53) [inline test/concat_test.dart]: I would expect this to be `secondType == 'single-subscription'` based on the test name?

File: test/concat_test.dart [hunk 5]
```diff
@@ -8,40 +8,121 @@ import 'package:test/test.dart';
...
+          });
+        }
+
+        test('closes stream after both inputs close', () async {
+          await first.close();
+          await second.close();
+          expect(isDone, true);
+        });
+
+        if (firstType == 'broadcast') {
```
Comment (jakemac53) [inline test/concat_test.dart]: it might be worth adding a `group` here with the name `firstType == 'broadcast'` or something

File: test/concat_test.dart [hunk 6]
```diff
@@ -8,40 +8,121 @@ import 'package:test/test.dart';
...
+              .listen(emittedValues.add, onError: errors.add, onDone: () {
+            isDone = true;
+          });
+        });
+
+        test('adds all values from both streams', () async {
+          first..add(1)..add(2);
+          await first.close();
+          await new Future(() {});
+          second..add(3)..add(4);
```
Comment (jakemac53) [inline test/concat_test.dart:59]: What if these are added before `first` closes? It seems like in the broadcast stream case you wouldn't catch any of those values?

File: lib/src/concat.dart (hunk 1 above)
Comment (natebosch) [reply lib/src/concat.dart]: The code paths around re-listening should only get hit if `first` is a broadcast stream, otherwise the controller would prevent us from ever getting here.

Our first listener can't hit this since `firstDone` won't ever be true. Subsequent listeners can only hit if `first` was a broadcast stream. We're not still listening to `first` since `firstDone` is true, which means we are (were) listening to `_next` so it's the only stream that `subscription` could be on.

File: test/concat_test.dart (hunk 2 above)
Comment (natebosch) [reply test/concat_test.dart:37]: I didn't think it was critical since the test that values come through prove we've listened when we should.

File: test/concat_test.dart (hunk 6 above)
Comment (natebosch) [reply test/concat_test.dart:59]: Correct. A lot of behavior in this transformer don't make much sense for broadcast streams, especially so for mixing them, but I thought it'd be best to handle it as well as possible. I can expand the doc comment to call some of this stuff out.

File: test/concat_test.dart (hunk 4 above)
Comment (natebosch) [reply test/concat_test.dart]: I might need to rename it. If the first stream is broadcast then a single-subscription second stream never gets canceled - only paused.

**Review by natebosch — COMMENTED** [review]
One thing I'm not sure of - If a single subscription stream is `concat`ed with a broadcast stream, then a cancel will cause it to `pause` - buffer values until it get's listened again.

Maybe instead I should replace the listener with something that ignores values instead?

File: test/concat_test.dart (hunk 6 above)
Comment (jakemac53) [reply test/concat_test.dart:59]: Technically if it is a broadcast stream it could start collecting things right away, and just not report them until the first stream is done right? I don't know if that is what users would expect or not.

**Comment by natebosch** [comment]
Come to think of it - what I did with `merge` was to use `asBroadcastStream()` for any single-subscription stuff that get's merged in... maybe that's what I should do here.

File: test/concat_test.dart (hunk 5 above)
Comment (natebosch) [reply test/concat_test.dart]: The current group names both types of streams so I don't think it would help the output. IMO the group would add a level of indent without much benefit.

File: lib/src/concat.dart (hunk 1 above)
Comment (natebosch) [reply lib/src/concat.dart]: Using `.asBroadcastStream()` allowed me to drop this block

File: test/concat_test.dart (hunk 6 above)
Comment (natebosch) [reply test/concat_test.dart:59]: Expanded the doc comment a bit.

Comment (natebosch) [reply test/concat_test.dart:59]: Somehow I missed the reply earlier.

I don't think it would be best to buffer - it's a potential memory leak. If we have to pick one of 2 bad behaviors I think missing earlier events is the less bad choice.

It's also possible to get that behavior manually by forcing back to single-subscription. Whereas it's not possible to get back to the other behavior if we were to buffer.

```dart
Stream<T> singleSubscriptionEarlyListen<T>(Stream<T> values) {
  var controller = new StreamController<T>();
  values.listen(controller.add, onError: controller.adError, onDone: controller.close);
  return controller.stream;
}
//...
someStream.transform(concat(singleSubscriptionEarlyListen(someBroadcast)));
```

**Comment by natebosch** [comment]
@jakemac53 any other thoughts?

**Review by jakemac53 — APPROVED** [review]
